src/components/cc-kv-console/cc-kv-console.js

```diff
@@ -0,0 +1,328 @@
+ * @cssprop {Color} --cc-kv-console-color-foreground-success - The foreground color of the result when command succeeded
+ * @cssprop {Color} --cc-kv-console-color-foreground-error - The foreground color of the result when command failed
+ */
+export class CcKvConsole extends LitElement {
+  static get properties() {
+    return {
+      state: { type: Object },
+    };
+  }
+
+  constructor() {
+    super();
+    /** @type {CcKvConsoleState} - The state of the component */
+    this.state = { type: 'idle', history: [] };
+
+    /** @type {Array<string>} */
+    this._cmdHistory = [];
+
+    /** @type {number|null} */
+    this._cmdHistoryIndex = null;
+
+    /** @type {HTMLInputElementRef} */
+    this._promptRef = createRef();
+  }
+
+  /**
+   * @param {HTMLInputKeyboardEvent} e
+   */
+  _onShellPromptKeyDown(e) {
+    e.stopPropagation();
+
+    if (e.key === 'Enter') {
+      e.preventDefault();
+      this._cmdHistoryIndex = null;
+      const commandLine = e.target.value.trim();
+
+      if (isStringEmpty(commandLine)) {
+        this.state = {
+          ...this.state,
+          history: [...this.state.history, { commandLine, result: [], success: true }],
+        };
+        dispatchCustomEvent(this, 'state-change', this.state);
+      } else if (commandLine.toLowerCase() === 'clear') {
+        this._pushCommandToHistory('clear');
+        this.state = {
+          ...this.state,
+          history: [],
+        };
+        dispatchCustomEvent(this, 'state-change', this.state);
+        e.target.value = '';
+      } else {
+        this._pushCommandToHistory(commandLine);
+        dispatchCustomEvent(this, 'send-command', commandLine);
+        e.target.value = '';
+      }
+    } else if (e.key === 'ArrowUp') {
+      e.preventDefault();
+
+      if (this._cmdHistory.length === 0) {
+        return;
+      }
+
+      if (this._cmdHistoryIndex == null) {
+        this._cmdHistoryIndex = this._cmdHistory.length - 1;
+      } else {
+        this._cmdHistoryIndex = Math.max(this._cmdHistoryIndex - 1, 0);
+      }
+
+      if (this._cmdHistoryIndex >= 0) {
+        this._promptRef.value.value = this._cmdHistory[this._cmdHistoryIndex];
+      }
+    } else if (e.key === 'ArrowDown') {
+      e.preventDefault();
+
+      if (this._cmdHistoryIndex != null) {
+        this._cmdHistoryIndex++;
+
+        if (this._cmdHistoryIndex < this._cmdHistory.length) {
+          this._promptRef.value.value = this._cmdHistory[this._cmdHistoryIndex];
+        } else {
+          this._cmdHistoryIndex = null;
+          this._promptRef.value.value = '';
+        }
+      }
+    }
+  }
+
+  /**
+   * @param {MouseEvent} e
+   */
+  _onContentMouseUp(e) {
+    if (e.target !== this._promptRef.value) {
+      setTimeout(() => {
+        if (isStringEmpty(document.getSelection()?.toString())) {
+          this._promptRef.value?.focus({ preventScroll: true });
+        }
+      }, 0);
```

Comment:
No. This is not related to Lit lifecycle. Here we want to prevent from the prompt to take focus when the user only selects some text inside the history.
I'll write a comment.

Reply:
We want to focus the prompt input when user clicks on the history panel.
But, in the meantime, we want to prevent the prompt from taking focus when the user only selects some text inside the history.
That's why we bind the logic on a `mouseup` event and check for selection before trying to focus the prompt input.

Note that the `document.getSelection()` is empty during the `mouseup` event.
That's why we need to postpone the logic inside a `setTimeout`.